### PR TITLE
Support rebuilds, add pre-commit

### LINKED ISSUE #18: Properly support rebuilds
On running sphinx-build for a 2nd time, you currently get warnings:

WARNING: (broken) content-types/markdown.html redirects to file-types/markdown.html but /Users/chrisjsewell/Documents/GitHub/jupyter-book/docs/_build/html/content-types/markdown.html already exists!

Possible approaches:

1. keep track of written redirect files in the sphinx `env`
2. Add an "identifier" comment at the top of the redirect files, then check for it

### PR DESCRIPTION
**Pull request by chrisjsewell**
Fixes #18

### CODE REVIEW

**Comment by Daltz333** [comment]
@chrisjsewell I'm aiming to push a release tomorrow night around 10:30PM EST.

**Comment by chrisjsewell** [comment]
> I'm aiming to push a release tomorrow night around 10:30PM EST.

Cool cheers

If you don't mind I added a [pre-commit](https://pre-commit.com/) config to the repo, and run the black CI *via* that (there's some other commented out checks you might consider adding)

**Comment by chrisjsewell** [comment]
I'm also running the tests *via* [tox](https://tox.readthedocs.io).
I can add the tox.ini if you want?
Basically both pre-commit and tox and super useful for not having to manually set up a development environment; you just run either and it will do it for you

**Comment by TheTripleV** [comment]
Sure! I'm open to the tox file.

**Comment by chrisjsewell** [comment]
> Could you add a dirhtml test case

Done 👍 

I figured out something a bit annoying; sphinx doesn't clean the build folder in-between each test.
Hence, the addition of the `shutil.rmtree` at the start of these rebuild tests

**Comment by TheTripleV** [comment]
Yeah I just noticed that while working on the other open pr. I'll make an issue as we should clean the build folder before every test.

File: tests/test_ext.py [hunk 1]
```diff
@@ -210,6 +221,16 @@ def test_simple(self, app: Sphinx, ensure_redirect):
         assert app.statuscode == 0
         ensure_redirect("another/index.html", "index.html")
 
+    @pytest.mark.sphinx("html", testroot="simple", freshenv=False)
```
Comment (TheTripleV) [inline tests/test_ext.py:224]:
```suggestion
    @pytest.mark.sphinx("dirhtml", testroot="simple", freshenv=False)
```

File: tests/test_ext.py [hunk 2]
```diff
@@ -210,6 +221,16 @@ def test_simple(self, app: Sphinx, ensure_redirect):
...
+    def test_simple_rebuild(self, app: Sphinx, ensure_redirect):
+        if Path(app.outdir).exists():
+            shutil.rmtree(Path(app.outdir))
+        app.build()
+        assert app.statuscode == 0
+        app.build()
+        assert app.statuscode == 0
+        ensure_redirect("another.html", "index.html")
```
Comment (TheTripleV) [inline tests/test_ext.py:232]:
```suggestion
        ensure_redirect("another/index.html", "index.html")
```

**Comment by chrisjsewell** [comment]
> Yeah I just noticed that while working on the other open pr. 

Yeh I guess just make a new fixture:

```python
@pytest.fixture()
def app_clean(app):
    if Path(app.outdir).exists():
      shutil.rmtree(Path(app.outdir))
    yield app
```

(or clean after)

**Comment by chrisjsewell** [comment]
> The dirhtml test is using the html builder.

Ooops my bad 🤦

**Comment by TheTripleV** [comment]
LGTM. This will get merged in tomorrow.

**Comment by chrisjsewell** [comment]
Brilliant, thanks for the great extension 😄

**Pull request merged**